### PR TITLE
Add FWaaS_V2

### LINKED ISSUE #115: feature request: support for FWaaS v2.0
Hi there,

my OpenStack provider only supports FWaaS v2.0, the v1.0 endpoint is disabled. Are there any plans already to add support for FWaaS v2.0 since v1.0 is flagged as deprecated?

Cheers,
Christian

### PR DESCRIPTION
**Pull request by Koodt**
Hi there!

Add FWaaS_V2 

Close 1/3 part of #115

Add functional-fwaas_v2 workflow.
Add fw_rule_v2 data source, resource, import, acc tests.
Add fw_rule_v2 resource and data sources to DataSourcesMap.
Add fw_rule_v2 website docs.
Add fw_rule_v2 to openstack.erb.

Bump Gophercloud version to 1.4.0.
Update go.sum.

### CODE REVIEW

**Comment by nikParasyr** [comment]
@Koodt Thanks for this.
CI is not working properly

This is a huge PR (which should have been broken down in multiple PRs). I will have time to review it in a couple of weeks

**Comment by Koodt** [comment]
Groups and policies was removed 
Keep only rule's doc, resources, datasource's, tests, CI workflow and golang bump

File: .github/workflows/functional-fwaas_v2.yml [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+concurrency:
+  group: ${{ github.workflow }}-${{ github.ref }}
+  cancel-in-progress: true
+on:
+  pull_request:
+    paths:
+      - '.github/workflows/functional-fwaas.yml'
+      - '**_fw_**'
+      - '!**v1**'
+      - 'CHANGELOG.md'
```
Comment (nikParasyr) [inline .github/workflows/functional-fwaas_v2.yml:12]: Add:
```
      - 'scripts/*'
```

File: .github/workflows/functional-fwaas_v2.yml [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+        name: ["master"]
+        openstack_version: ["master"]
+        ubuntu_version: ["22.04"]
+        include:
+          - name: "antelope"
+            openstack_version: "stable/2023.1"
+            ubuntu_version: "22.04"
+          - name: "zed"
+            openstack_version: "stable/zed"
+            ubuntu_version: "22.04"
```
Comment (nikParasyr) [inline .github/workflows/functional-fwaas_v2.yml:29]: Why are yoga and xena not added? We typically try and test master + latest 4 releases

File: openstack/data_source_openstack_fw_rule_v2.go [hunk 3]
```diff
@@ -0,0 +1,215 @@
...
+				Optional: true,
+				Computed: true,
+			},
+
+			"description": {
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"protocol": {
```
Comment (nikParasyr) [inline openstack/data_source_openstack_fw_rule_v2.go:47]: Can you add a validateFunc similar to [here](https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/main/openstack/resource_openstack_lb_l7policy_v2.go#L64).

Also on your code later on I also see protocol `any`, but this is not the case based on the api docs. Am i missing something?

File: openstack/data_source_openstack_fw_rule_v2.go [hunk 4]
```diff
@@ -0,0 +1,215 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				StateFunc: func(val any) string {
+					return strings.ToLower(val.(string))
+				},
+			},
+
+			"action": {
```
Comment (nikParasyr) [inline openstack/data_source_openstack_fw_rule_v2.go:55]: ValidateFunc here as well

File: openstack/data_source_openstack_fw_rule_v2.go [hunk 5]
```diff
@@ -0,0 +1,215 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+			},
+
+			"ip_version": {
```
Comment (nikParasyr) [inline openstack/data_source_openstack_fw_rule_v2.go:60]: ValidateFunc

File: openstack/data_source_openstack_fw_rule_v2.go [hunk 6]
```diff
@@ -0,0 +1,215 @@
...
+	if v, ok := d.GetOk("description"); ok {
+		listOpts.Description = v.(string)
+	}
+
+	if v, ok := d.GetOk("tenant_id"); ok {
+		listOpts.TenantID = v.(string)
+	}
+
+	if v, ok := d.GetOk("protocol"); ok {
+		listOpts.Protocol = expandFWRuleV2Protocol(v.(string))
```
Comment (nikParasyr) [inline openstack/data_source_openstack_fw_rule_v2.go:135]: Is expandFWRuleV2Protocol needed? Wouldnt an approach like [this](https://github.com/terraform-provider-openstack/terraform-provider-openstack/blob/main/openstack/resource_openstack_lb_l7policy_v2.go#L134) not work here as well?

File: openstack/data_source_openstack_fw_rule_v2.go [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+	}
+
+	if v, ok := d.GetOk("action"); ok {
+		listOpts.Action = expandFWRuleV2Action(v.(string))
```
Comment (nikParasyr) [inline openstack/data_source_openstack_fw_rule_v2.go:139]: Same as above

File: openstack/fw_rule_v2.go [hunk 8]
```diff
@@ -0,0 +1,50 @@
+package openstack
+
+import (
+	"strings"
+
+	"github.com/gophercloud/gophercloud"
+	"github.com/gophercloud/gophercloud/openstack/networking/v2/extensions/fwaas_v2/rules"
+)
+
+func expandFWRuleV2Action(a string) rules.Action {
```
Comment (nikParasyr) [inline openstack/fw_rule_v2.go:10]: Check my comment above. I think these functions are not necessary

File: openstack/resource_openstack_fw_rule_v2.go [hunk 9]
```diff
@@ -0,0 +1,307 @@
...
+				Optional: true,
+			},
+
+			"tenant_id": {
+				Type:     schema.TypeString,
+				Optional: true,
+				ForceNew: true,
+			},
+
+			"protocol": {
```
Comment (nikParasyr) [inline openstack/resource_openstack_fw_rule_v2.go:49]: Same as the datasource. Please include a ValidateFunc and remove the StateFunc when it doesn't make sense

File: openstack/resource_openstack_fw_rule_v2.go [hunk 10]
```diff
@@ -0,0 +1,307 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				StateFunc: func(val any) string {
+					return strings.ToLower(val.(string))
+				},
+				Default: "any",
+			},
+
+			"action": {
```
Comment (nikParasyr) [inline openstack/resource_openstack_fw_rule_v2.go:58]: same

File: openstack/resource_openstack_fw_rule_v2.go [hunk 11]
```diff
@@ -0,0 +1,307 @@
...
+				Type:     schema.TypeString,
+				Optional: true,
+				StateFunc: func(val any) string {
+					return strings.ToLower(val.(string))
+				},
+				Default: "deny",
+			},
+
+			"ip_version": {
```
Comment (nikParasyr) [inline openstack/resource_openstack_fw_rule_v2.go:67]: same

File: openstack/resource_openstack_fw_rule_v2.go [hunk 12]
```diff
@@ -0,0 +1,307 @@
...
+	if err != nil {
+		return diag.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	shared := d.Get("shared").(bool)
+	enabled := d.Get("enabled").(bool)
+	ruleCreateOpts := rules.CreateOpts{
+		Name:                 d.Get("name").(string),
+		Description:          d.Get("description").(string),
+		Protocol:             expandFWRuleV2Protocol(d.Get("protocol").(string)),
```
Comment (nikParasyr) [inline openstack/resource_openstack_fw_rule_v2.go:122]: As mentioned, these functions are probably not necessary

File: openstack/resource_openstack_fw_rule_v2.go [hunk 13]
```diff
@@ -0,0 +1,307 @@
...
+	if err != nil {
+		return diag.Errorf("Error creating OpenStack networking client: %s", err)
+	}
+
+	rule, err := rules.Get(networkingClient, d.Id()).Extract()
+	if err != nil {
+		return diag.FromErr(CheckDeleted(d, err, "Error retrieving openstack_fw_rule_v2"))
+	}
+
+	if len(rule.FirewallPolicyID) > 0 {
```
Comment (nikParasyr) [inline openstack/resource_openstack_fw_rule_v2.go:291]: Wondering whether this should be here. In general we let tf take care of dependencies and kill/delete related resources accordingly. If a resource is managed outside of tf then we don't take action.

In this case, tf should, from the depedency graph, delete the policy first and then move to the rules, thus i dont think we need to have this check. What is your opinion?

**Review by nikParasyr — COMMENTED** [review]
@Koodt Great PR! 
I made some comments about things to fix. Most of them are "nit"/easy to do. Let me know if you need any clarifications somewhere

File: .github/workflows/functional-fwaas_v2.yml (hunk 2 above)
Comment (Koodt) [reply .github/workflows/functional-fwaas_v2.yml:29]: FWaaS_v2 was deprecated in this cycles, but was revived in Zed.
Openstack releases repo on Opendev hasn't FWaaS_V2 yaml.
Neutron-fwaas repo on Opendev has only ussuri/zed/antilope branches.

File: openstack/data_source_openstack_fw_rule_v2.go (hunk 3 above)
Comment (Koodt) [reply openstack/data_source_openstack_fw_rule_v2.go:47]: Good point about validate.

The Networking client requires a protocol value. If the rule is protocol agnostic, you can use the `any` value.

File: openstack/resource_openstack_fw_rule_v2.go (hunk 13 above)
Comment (Koodt) [reply openstack/resource_openstack_fw_rule_v2.go:291]: Understood.
If we start managing the association from a policy resource, this check will not be needed.

**Review by nikParasyr — APPROVED** [review]
@Koodt thanks you for this PR and all the work. Also for the patience for the reviews. 

This looks good to me.

You can create the rest of the PRs for the fw_v2 resources (incorporate the feedback from this PR please -- such as ValidateFunc etc). I think there should be 2 resources and 2 data_sources left. so you can create 1 PR for each one of them ( 4 in total ), and ill try to review them as soon as possible and I think ill make a release after they merged ( together with a couple PRs that are already open)

**Pull request merged**